### PR TITLE
Shooter Subsystem - Issue #45

### LINKED ISSUE #20: Prototyping: Operator Control Rotate/Aim Robot

### LINKED ISSUE #45: Shooter Subsystem
Production software for the shooter subsystem in the robot code, including:
- interface definitions
- commands
- PID
- etc..
- tested independently
- merged and integrated with 2020 robot code in github
- ready for integration

### LINKED ISSUE #88: Refactor shooter subsystem to use NEO motor
D&F would like to switch from the 2x775 motor assembly to use a single Rev Robotics NEO brushless motor controlled by a Spark Max motor controller.

This will require the shooter subsystem to be modified. The commands and general structure don't need to change. The good news is that the Spark Max has a pretty good software API, examples on GitHub, and control of a single motor with a built-in encoder is quite easy. (Possibly even easier than what we were going to do with the 2x775 approach)

It's also quite comparable to the Talon SRX code from CTRE, so it should look fairly familiar.

See here for example code:

https://github.com/REVrobotics/SPARK-MAX-Examples/tree/master/Java/Velocity%20Closed%20Loop%20Control

You will also need to download the Spark Max libraries to your computer, change the build.gradle to point to those new vendor libraries.

To test, we will either get a standalone NEO and Spark Max to connect to one of the robots, or we will switch out the existing CIM motor in the prototype fuel shooter.

### PR DESCRIPTION
**Pull request by LaraLim**
ShooterSubsystem, Config, and the SpinUpShooter command currently use some protobot-specific values, like the CAN ID constant and PID values. This pull request will not be merged until the code is properly tested on protobot. Parts of the code that are relevant only to protobot will then be changed or removed.

### CODE REVIEW

**Comment by georgetzavelas** [comment]
Looks like you need a rebase

File: src/main/java/frc/robot/RobotContainer.java [hunk 1]
```diff
@@ -64,6 +66,10 @@ private void configureButtonBindings() {
     // Instantiate the intake command and bind it
     intakeCommand = new OperatorIntakeCommand();
     new JoystickButton(driverStick, XboxController.Button.kBumperLeft.value).whenHeld(intakeCommand);
+
+    // Instantiate the shooter ramping command and bind it
+    rampShooterCommand = new SpinUpShooter();
+    new JoystickButton(driverStick, XboxController.Button.kA.value).whenPressed(rampShooterCommand);
```
Comment (kevlam2706) [inline src/main/java/frc/robot/RobotContainer.java]: How will this work - will it hold the shooter flywheel at speed while the button is pressed down?

File: src/main/java/frc/robot/commands/SpinUpShooter.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    addRequirements(shooterSubsystem);
+  }
+
+  @Override
+  public void initialize() {
+  }
+
+  @Override
+  public void execute() {
+    // Hardcoded RPM for testing
```
Comment (kevlam2706) [inline src/main/java/frc/robot/commands/SpinUpShooter.java]: maybe set a variable in the command that we temporarily set to 2000 RPM for testing, but eventually will be determined by a call to some kind of formula. Then refer to the variable on line 22 and 24.

File: src/main/java/frc/robot/config/FluidConstant.java [hunk 3]
```diff
@@ -76,14 +76,6 @@ public T getValue() {
         return this.value;
     }
     
-    /**
-     * Get method for implementing the Supplier interface
-     * @return The current value of this constant
-     */
-    public T get() {
-        return this.value;
```
Comment (kevlam2706) [inline src/main/java/frc/robot/config/FluidConstant.java]: I notice a lot of people who have this deletion in their code reviews. Any idea what's going on?

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  // Singleton instance for ShooterSubsystem
+  private static final ShooterSubsystem INSTANCE_SHOOTER = new ShooterSubsystem();
+
+  private WPI_TalonSRX m_shooter;
+
+  // Indicates use of the primary PID loop rather than cascaded ones
+  int kPIDLoopIDX = 0;
+
+	 // Set to zero to skip waiting for confirmation, set to nonzero to wait and
+	 // report to DS if action fails.
```
Comment (kevlam2706) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java]: watch out for your indenting..

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    if (Config.SHOOTER_MOTOR != -1){
+      m_shooter = new WPI_TalonSRX(Config.SHOOTER_MOTOR);
+    }
+
+    // Factory Default to prevent unexpected behaviour
+    m_shooter.configFactoryDefault();
+
+    // Config sensor used for PID control
+    m_shooter.configSelectedFeedbackSensor(FeedbackDevice.CTRE_MagEncoder_Relative, 
+    kPIDLoopIDX, kTimeoutMs);
```
Comment (kevlam2706) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java]: eventually add error checking here in case the encoder isn't connected, we don't want to try running a closed loop mode

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    m_shooter.configFactoryDefault();
+
+    // Config sensor used for PID control
+    m_shooter.configSelectedFeedbackSensor(FeedbackDevice.CTRE_MagEncoder_Relative, 
+    kPIDLoopIDX, kTimeoutMs); 
+
+    // Invert the phase of the sensor so positive output yields positive change
+    m_shooter.setSensorPhase(true);
+    
+		// Config the peak and nominal outputs
```
Comment (kevlam2706) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java]: indenting

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+      return INSTANCE_SHOOTER;
+  }
+
+  /**
+   * Return the motor velocity measured by the encoder
+   */
+  public double getVelocity(){
+    double encoderVelocity = m_shooter.getSelectedSensorVelocity();
+    // Add the ability to print to SmartDashboard?
+    return encoderVelocity;
```
Comment (kevlam2706) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java]: what are the units here? can we turn them back into RPM?

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+  public void checkRPM(int targetRPM){
+
+    // Calculate RPM based on the encoder reading
+    double calculatedRPM = (m_shooter.getSelectedSensorVelocity() * 600) / 4096;
+
+    // Verify that the motor is running at the target RPM
+    if (calculatedRPM < (targetRPM + 50) && calculatedRPM > (targetRPM - 50)){
+      System.out.println("calculatedRPM is within 50 units of targetRPM");
+      // I have no idea if +/- 50 tolerance around the RPM is accurate enough
+      // Test and change, add setting conditions
```
Comment (kevlam2706) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java:96]: Suggest that we put the tolerance value (50) in a class variable so we can easily change it. Then just refer to it here.

Also, this method should be public boolean instead of public void, so that it can return true if at speed or false if not.

**Review by kevlam2706 — APPROVED** [review]
Looks great! Some minor formatting fixes and other suggestions.

File: src/main/java/frc/robot/commands/SpinUpShooter.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    shooterSubsystem.setRPM(2000);
```
Comment (georgetzavelas) [inline src/main/java/frc/robot/commands/SpinUpShooter.java]: Can we make the number into a constant?

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+import com.ctre.phoenix.motorcontrol.ControlMode;
+import com.ctre.phoenix.motorcontrol.FeedbackDevice;
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+import frc.robot.config.Config;
+
+public class ShooterSubsystem extends SubsystemBase {
+
+  // Singleton instance for ShooterSubsystem
+  private static final ShooterSubsystem INSTANCE_SHOOTER = new ShooterSubsystem();
```
Comment (georgetzavelas) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java]: Instead of initializing here, consider using a Singleton pattern that lazy loads instead of relying on static loading like this: https://github.com/georgetzavelas/guava-examples/blob/master/src/com/design/pattern/creational/singleton/Singleton.java

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    // Add the ability to print to SmartDashboard?
+    return encoderVelocity;
+  }
+
+  /**
+   * Set the target RPM to ramp up to. Will eventually get this value 
+   * based on distance from power port or other factors.
+   */
+  public void setRPM(int targetRPM){
+    velocityModeUnits = targetRPM * 4096 / 600;
```
Comment (georgetzavelas) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java]: Can we make these numbers into constants?

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+   * Check the actual RPM and compare it with targetRPM
+   * to verify that the shooter is up to necessary speed to fire.
+   */
+  public void checkRPM(int targetRPM){
+
+    // Calculate RPM based on the encoder reading
+    double calculatedRPM = (m_shooter.getSelectedSensorVelocity() * 600) / 4096;
+
+    // Verify that the motor is running at the target RPM
+    if (calculatedRPM < (targetRPM + 50) && calculatedRPM > (targetRPM - 50)){
```
Comment (georgetzavelas) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java]: Can we make these numbers into constants?

**Review by georgetzavelas — COMMENTED** [review]
Awesome progress! Some comments for your consideration

File: src/main/java/frc/robot/RobotContainer.java (hunk 1 above)
Comment (LaraLim) [reply src/main/java/frc/robot/RobotContainer.java]: The command will be scheduled when the button goes from inactive to active. The button doesn't have to be held continuously for the shooter to ramp up. The command can be rebound to another trigger based on drive team preference or anything that makes more sense.

File: src/main/java/frc/robot/config/FluidConstant.java (hunk 3 above)
Comment (nrgill28) [inline src/main/java/frc/robot/config/FluidConstant.java]: This issue is fixed in master. Can you revert the changes to this file so we don't break it again?

**Review by nrgill28 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java (hunk 7 above)
Comment (LaraLim) [reply src/main/java/frc/robot/subsystems/ShooterSubsystem.java]: The units read off the encoder are position change per 100ms. CheckRPM() uses it to calculate the current RPM when the shooter is running and compares it with targetRPM.

File: src/main/java/frc/robot/config/FluidConstant.java (hunk 3 above)
Comment (georgetzavelas) [reply src/main/java/frc/robot/config/FluidConstant.java]: Rebasing should simply remove this problem, no need to explicitly revert anything.

**Review by Zinka010 — APPROVED** [review]
Seems good for now considering the changes

File: src/main/java/frc/robot/subsystems/DriveBase.java [hunk 13]
```diff
@@ -54,7 +54,7 @@ private DriveBase() {
         talon5plyboy = new WPI_TalonSRX(Config.TALON_5_PLYBOY);
 
         robotDriveBase = new DifferentialDrive(leftFrontTalon, rightFrontTalon);
-
+        _pidgey = new PigeonIMU (Config.robotSpecific(null, null, null, leftFrontTalon, leftRearTalon, talon5plyboy));
```
Comment (georgetzavelas) [inline src/main/java/frc/robot/subsystems/DriveBase.java]: Is this needed as I don't see it being used anywhere?

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+import com.revrobotics.ControlType;
+import com.revrobotics.CANSparkMaxLowLevel.MotorType;
+
+public class ShooterSubsystem extends SubsystemBase {
+
+  private CANSparkMax m_shooter;
+  private CANPIDController m_pidController;
+  private CANEncoder m_encoder;
+
+  double kF = 0.00018;
```
Comment (georgetzavelas) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java]: Can you make these constants? Also what are the values for, e.g. kF == SHOOTER_FUDGE_FACTOR?

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+   */
+  public void setRPM(int inputRPM){
+    setpointRPM = inputRPM;
+  }
+
+  /**
+   * Return the motor velocity measured by the encoder
+   */
+  public double getRPM(){
+    double encoderRPM = m_encoder.getVelocity();
```
Comment (georgetzavelas) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java]: nitpick: you don't need the temporary variable here.

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+   * Returns the singleton instance for the ShooterSubsystem
+   */
+  public static ShooterSubsystem getInstance() {
+    return ShooterHolder.INSTANCE_SHOOTER;
+  }
+
+  /**
+   * Set the target RPM to ramp up to.
+   */
+  public void setRPM(int inputRPM){
```
Comment (georgetzavelas) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java]: How about `setTargetRPM`?

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+  public double getRPM(){
+    double encoderRPM = m_encoder.getVelocity();
+    return encoderRPM;
+  }
+
+  /**
+   * Check the actual RPM and compare it with targetRPM
+   * to verify that the shooter is up to necessary speed to fire.
+   */
+  public boolean checkRPM(int inputRPM){
```
Comment (georgetzavelas) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java]: Do you expect `inputRPM` to be different than what is currently stored for `setpointRPM`? If not then it would may make sense to simply use that without needing to pass in a variable. Also this function feels more like a `isAtTargetRPM` method

File: src/main/java/frc/robot/subsystems/ShooterSubsystem.java [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+    double encoderRPM = m_encoder.getVelocity();
+    return (encoderRPM < (inputRPM + RPM_TOLERANCE) && encoderRPM > (inputRPM - RPM_TOLERANCE));
```
Comment (georgetzavelas) [inline src/main/java/frc/robot/subsystems/ShooterSubsystem.java]: How about
```
return (Math.abs(inputRPM - encoderRPM) < RPM_TOLERANCE)
```

**Review by georgetzavelas — COMMENTED** [review]
This is coming along great Lara! I left comments in one Subsystem that applies to the other.

File: src/main/java/frc/robot/subsystems/DriveBase.java (hunk 13 above)
Comment (Lockscout) [reply src/main/java/frc/robot/subsystems/DriveBase.java]: This is code that I have written that ended up on her branch. The code here is outdated and if it were to be put into master, it would break master. Currently communicating to fix it. But yeah this code is needed, as its used in drivetrainPID stuff.

Comment (Lockscout) [reply src/main/java/frc/robot/subsystems/DriveBase.java]: Ok it isnt used, like you said. I misunderstood the issue. That line just needs to be deleted. Sorry.

**Review by nrgill28 — APPROVED** [review]

**Review by georgetzavelas — APPROVED** [review]

**Comment by Lockscout** [comment]
Alright, so here's what we've concluded after testing today.

The shooter ramping, and pid works great! Of course it still needs a bit of tuning, but overall it works good.

There is one issue with it. After you turn on the robot, or upload new code to the roboRIO, The target speed is always set to 2000 (as that's what the default is) and causes it to ramp up to that rpm when its enabled, until you hit the button once to set the target to 0. What might need to be done is when the robot is initializing, set the target to 0.
